### PR TITLE
Movie Editions Feature

### PR DESCRIPTION
**Pull request by ccope**
## Description
I'm reviving #14972, I'm hoping this can make it into Kodi 20. Huge thanks to @xodidox for getting this so close to the finish line!

Remaining TODOs from the previous PR:
- [ ] Bikeshed on the name of the db and in the UI
- [ ] Decide whether to include any edition labels by default, and how
- [ ] I have a question about the SQL statements in CVideoDatabase::GetMovieId

Copying most of the rest of this template from the original PR.

### Added movie version feature

With this feature, it's possible to connect more than one video file with one movie entry in library. For example, Kodi will support both STANDARD version and EXTENDED version of a movie live under one movie item. A context menu item "Manage movie version" under "Manage" menu item has been added to the GUI. A "Version" button has also been added to the video info dialog for movie, which allows user change the default video file for the movie. The change will be saved to DB.

Step by step instruction for usage:
https://forum.kodi.tv/showthread.php?tid=337992

## Motivation and context
Adding movie version support to Kodi has been asked many times in forum and I personally also want this feature for a long time

## How has this been tested?
This was last tested before I rebased it. I need to get my Kodi test environment working again, and I'll see if anyone else in the forum thread can test as well. I just want to get the bikeshedding done in parallel.

## What is the effect on users?
This will allow users to add extra release types (Director's Cut, Extended Edition, RiffTrax, etc), import extra copies of movies with those release types, and choose which type of the film to play.

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [X] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [X] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: xbmc/Application.cpp [hunk 1]
```diff
@@ -2502,8 +2503,60 @@ bool CApplication::PlayStack(CFileItem& item, bool bRestart)
   return PlayFile(selectedStackPart, "", true);
 }
 
+void CApplication::SelectMovieVersion(CFileItem& item)
```
Comment (enen92) [inline xbmc/Application.cpp:2506]: I don't think dumping even more GUI stuff into CApplication is the way to go unfortunately. We need some sort of mechanism to bridge this into GUI via messaging, I am not sure if we already have any pattern like this in the codebase.

File: xbmc/interfaces/legacy/InfoTagVideo.h [hunk 2]
```diff
@@ -2297,6 +2297,23 @@ namespace XBMCAddon
...
+      ///
+      /// \ingroup python_InfoTagVideo
+      /// @brief \python_func{ setVersion(version) }
+      /// Set the edition of the video item.
+      ///
+      /// @param title              string - Version.
+      ///
+      ///
+      ///-----------------------------------------------------------------------
+      /// @python_v21 New function added.
```
Comment (enen92) [inline xbmc/interfaces/legacy/InfoTagVideo.h:2310]: v20? we are not on the finish line to consider this only for v21 IMHO

File: xbmc/video/VideoInfoScanner.cpp [hunk 3]
```diff
@@ -20,6 +20,8 @@
 #include "cores/VideoPlayer/DVDFileInfo.h"
 #include "dialogs/GUIDialogExtendedProgressBar.h"
 #include "dialogs/GUIDialogProgress.h"
+#include "dialogs/GUIDialogSelect.h"
```
Comment (enen92) [inline xbmc/video/VideoInfoScanner.cpp:23]: another component that was not supposed to be filled with gui includes

File: xbmc/Application.cpp (hunk 1 above)
Comment (ccope) [reply xbmc/Application.cpp:2506]: I don't disagree, though I'm not very familiar with this codebase and my C++ skills are mediocre at best. The only thing I can think of is maybe mimicking the resume/restart logic? But after doing some skimming it looks like maybe too much of that is stashed in the PlayerCore, and this definitely doesn't belong there either.

File: xbmc/video/windows/GUIWindowVideoBase.cpp [hunk 4]
```diff
@@ -1186,7 +1186,7 @@ bool CGUIWindowVideoBase::OnPlayAndQueueMedia(const CFileItemPtr& item, const st
 
 void CGUIWindowVideoBase::PlayMovie(const CFileItem *item, const std::string &player)
 {
-  if(m_thumbLoader.IsLoading())
+  if (m_thumbLoader.IsLoading())
```
Comment (garbear) [inline xbmc/video/windows/GUIWindowVideoBase.cpp:1189]: Unrelated change (but still welcome). Don't worry about this now, but as the PR progresses you'll incorporate feed and combine the commits into a clean history, and at that point you'll want to split this out into a second commit.

Keep an eye on those "lines added" and "lines deleted" metrics. The main commit should ideally have 0 lines deleted. This indicates to me that no existing code is touched, making the feature much safer and more desirable.

**Comment by garbear** [comment]
I'll do a thorough review when the PR is marked as ready status. I really like this feature though. Question, how do you scrape the movie edition? My current approaches uses a script that looks for e.g. "directors_cut" in the filename and modifies the nfo file to modify the title shown in the GUI

Screen Shot 2022-05-29 at 3 23 45 PM

However, this has the downside that if I've ever renamed a file in my ripping history, then I could lose the edition modifier.

**Comment by ccope** [comment]
> How do you scrape the movie edition?

Currently, there is no automatic scraping. Creating edition types and categorizing files is currently a manual process left to the user. IMO this feature is most useful for cases where a user would want to keep multiple versions of the same film in their library. I think there are a few versions that might be broadly used, but I think in most cases users won't keep multiple versions of those films anyway. There are also a lot of versions that are one-off's (typically special cuts) that will never be scrape-able unless some kind of standard metadata field is created. I don't think I'd use this for different resolutions or 3D since that should already be visible from the codec metadata.

**Comment by garbear** [comment]
In a library of 1626 movies, these are the 115 marked editions I have:

```
[30th Anniversary Edition]
[35mm Edition Blu-ray Disc]
[35mm Edition]
[3D Diamond Edition]
[3D Extended Cut]
[3D]
[3D]
[3D]
[3D]
[3D]
[3D]
[3D]
[3D]
[3D]
[4K]
[Despecialized Edition]
[Despecialized Edition]
[Despecialized Edition]
[Diamond Edition]
[Director's Cut]
[Director's Cut]
[Director's Cut]
[Director's Cut]
[Director's Cut]
[Director's Cut]
[Director's Cut]
[Director's Cut]
[Director's Cut]
[Director's Cut]
[Director's Cut]
[Director's Cut]
[Director's Cut]
[Director's Cut]
[Extended Collector's Edition]
[Extended Cut]
[Extended Cut]
[Extended Cut]
[Extended Cut]
[Extended Edition]
[Extended Edition]
[Extended Edition]
[Extended Edition]
[Extended Edition]
[Extended Edition]
[Extended Edition]
[Extended Edition]
[Extended Edition]
[Extended Edition]
[Extended Edition]
[Extended Edition]
[Extended Edition]
[Extended Edition]
[Extended Edition]
[Extended Edition]
[Extended Edition]
[Extended Edition]
[Extended Fan Cut]
[Final Cut]
[Recut, Extended, Unrated]
[Remastered]
[Remastered]
[Remastered]
[Remastered]
[Remastered]
[Remastered]
[Special Assembly Cut]
[Special Collector's Edition]
[Special Edition]
[Special Extended Edition]
[Ultimate Cut]
[Ultimate Director's Cut]
[Uncut Extended]
[Uncut]
[Uncut]
[Unrated and Raw]
[Unrated Director's Cut]
[Unrated Director's Cut]
[Unrated Director's Cut]
[Unrated]
[Unrated]
[Unrated]
[Unrated]
[Unrated]
[Unrated]
[Unrated]
[Unrated]
[Unrated]
[Unrated]
[Unrated]
[Unrated]
[Unrated]
[Unrated]
[Unrated]
[Unrated]
[Unrated]
[Unrated]
[Unrated]
[Unrated]
[Unrated]
[Unrated]
[Unrated]
[Unrated]
[Unrated]
[Unrated]
[Unrated]
[Unrated]
[Unrated]
[Unrated]
[Unrated]
[Unrated]
[Unrated]
[Unrated]
[Unrated]
[Unrated]
[Unrated]
```

File: xbmc/FileItem.cpp [hunk 5]
```diff
@@ -1706,6 +1706,7 @@ void CFileItem::UpdateInfo(const CFileItem &item, bool replaceLabels /*=true*/)
       m_videoInfoTag = new CVideoInfoTag;
     }
 
+    SetDynPath(item.GetDynPath());
```
Comment (enen92) [inline xbmc/FileItem.cpp:1709]: please also check if this is needed. GetDynpath will fallback to the path if dynpath is not set. I don't see a reason why you need to set a dynpath to the already defined dynpath. The only reason I see is to force the path to be the dynpath in cases where the dynpath is not set. But then, getdynpath will always fallback to the path which makes me question the need here

**Comment by KarellenX** [comment]
After the rebase, if someone could kick off a build, I can then test this.
Thanks.

**Comment by peterhoeg** [comment]
Both jellyfin and emby support a scheme, where the resolution/version is stored as part of the file name: https://jellyfin.org/docs/general/server/media/movies.html#multiple-versions-of-a-movie

Example:
```
Movies
└── Best_Movie_Ever (2019)
    ├── Best_Movie_Ever (2019) - [1080P].mp4
    ├── Best_Movie_Ever (2019) - [720P].mp4
    └── Best_Movie_Ever (2019) - [Directors Cut].mp4
```

In order for kodi to co-exist nicely with them, it might be an idea to support a similar scheme. It's also an awful lot nicer to simply name files a certain way rather than having to manage the versions as part of the kodi UI.

**Comment by garbear** [comment]
You make a good point. The first thing is to gather all existing standards for editions. Then we choose the one that makes the most sense, or define our own.

I personally don't like the example, because I have a requirement - the file name has to match the folder name. Here is how my library is laid out:

```
Movies
└── Best_Movie_Ever (2019) [1080P]
    └── Best_Movie_Ever (2019) [1080P].mp4
└── Best_Movie_Ever (2019) [720P]
    └── Best_Movie_Ever (2019) [720P].mp4
└── Best_Movie_Ever (2019) [Director's Cut]
    └── Best_Movie_Ever (2019) [Director's Cut].mp4
```

**Comment by peterhoeg** [comment]
There are 2 different things at play here:

1. Same movie, different resolutions/audio
   
2. Different editions of the same movie with possibly different release years
   
The use cases are different (and so are the requirements as a consequence). In case of the first one, you would probably only want a single entry for the movie, but have an option to choose the resolution on launch.

In case of the latter, you would want to see each version as a separate entry as they are technically (or at least can be) different movies.

**Comment by gmitch64** [comment]
I'm very interested in this update. My main use is going to be for different resolutions (1080p/2160p), but my main hope is that I can use it to properly handle my 3D movies.

Is this going to make it into Nexus? I don't see it on the list of features (but I may well have missed it of course).

**Comment by da-anda** [comment]
I don't think [720] or [1080p] would have to be an "edition", as we can get this info directly from the stream details. So we could automatically group those kind of videos without the necessity to add some special "edition" parsing logic for these. So I would suggest that whatever is inside `[ ]` will be treated as edition.

**Comment by peterhoeg** [comment]
Instead of having special handling to read the stream details which would in any case require you to have *some* mechanism to say "these files are actually all the same movie, but just different editions", wouldn't it be far simpler to only deal with edition through the square brackets (or whatever mechanism is made)? 

You could easily foresee that some people for the sake of keeping things easy on the home-front would label them as `Some Movie (1998) [For mobile]` and `Some Movie (1998) [Home Theater]` to distinguish between 720p and 2160p to avoid getting the inevitable "why is this stupid movie lagging on my phone" from a dear family member.

**Comment by peterhoeg** [comment]
Kodi really doesn't have to care about the editions other than "here's a bunch of files that all resolve to the same movie so prompt to choose edition on play". This (to me) seems like the most intuitive and simplest solution overall.

**Comment by hjk321** [comment]
> Wouldn't it be far simpler to only deal with edition through the square brackets (or whatever mechanism is made)?

I agree with these sentiments, as it allows for non-technical users to get multiple versions of movies without much hassle.

**Comment by KarellenX** [comment]
Looks like we missed this being merged into v20. 
It would be nice to get this into v21 as early as possible.

What else needs to be done before it can be merged? I am willing to test if that is needed.

**Pull request closed without merging**